### ISSUE TITLE
[Help wanted] How can we improve the documentation?

### ISSUE THREAD

**Issue by jmeistrich** [LegendApp/legend-state#148]
We just released a big update to the docs, with a lot of restructuring and updating to the current feature set in version 1.5.0. I think we still have a ways to go to make the docs really great, so feedback or PRs to improve the docs would be very much appreciated!

### The plan from here is
1. Update the examples with a toggle to show a full interactive example. I ***think*** it's useful to have some examples be short in order to focus on the concepts they're explaining, so they skip showing all imports and some extraneous details, but it would also be useful to see the full code be interactive. So I'm thinking a toggle between them would be good? Or should they just all be the full code and interactive?
2. Add more details about props and parameters to the API pages
3. Add more larger scale usage examples
4. Add a search feature?

### Questions
1. Is this new organization structure good? Should some things be rearranged?
2. Do you notice any features missing documentation?
3. Is there anything that still confuses you that should be explained better?
4. How else could we improve the docs?
5. Do you want to help improve the docs?

---

## Major changes in the new docs

You may notice some big changes in the React example code. Based on feedback that some things were confusing and new features in 1.3.0 - 1.5.0, we've updated the docs to encourage new patterns and will deprecate the old patterns in a later version.

### Restructured

We rearranged concepts to be introduced in what seems like a better order and combined some pages together. For React we made a stronger split between core features and fine-grained-reactivity features, because many teams want to stick to standard React patterns in their initial adoption.

### Deprecating direct render of observables

We're deprecating `enableLegendStateReact` in favor of wrapping in `Memo`. Many were confused about different aspects of direct rendering, it was hard to see at a glance where observables were rendering, and it was hard to search files for rendered observables. Plus, it involved a bit of tricking React internals which we want to avoid. So we're going to deprecate that and render observables by just wrapping them in `Memo`, which looks like this:

```jsx
<Memo>{state$.text}</Memo>
```

See https://legendapp.com/open-source/state/fine-grained-reactivity/#render-an-observableselector-directly for more details.

### use() to make observables reactive

Many expected `get()` to make components reactive, were confused about the observer HOC, or don't like HOCs at all. Plus, it's easy to forget to add observer and end up with non-reactive components. So we have a new `use()` function that can be enabled on all observables (which is just useSelector under the hood). And the `observer` HOC can now optimize all instances of `use()` to be grouped into one hook, so we're considering it more of an optimization or advanced feature than a basic feature.

```jsx
function Counter () {
  const count = state$.count.use()
  return <div>{count}</div>
}
```

See https://legendapp.com/open-source/state/react-introduction/ for more details.

### `Reactive` components replace `Legend` components

We did not like having different `Legend` components per platform at different import paths, and now that we are using a configuration pattern it's easy to configure Reactive with the components for your platform. See https://legendapp.com/open-source/state/fine-grained-reactivity/#reactive-components for more details.

## What do you think?

If you have any thoughts or feedback on the above questions or want to help with the docs, please jump in 🤟

**Comment by chakrihacker** [comment]
If enableLegendStateForReact is removed wouldn’t it be a big refactor for existing code base??

Assuming most of the components are wrapped with observer HOC

**Comment by jmeistrich** [comment]
Removing enableLegendStateForReact wouldn't affect observer - it only enables direct rendering of observables. Unfortunately it may be an annoying refactor to make, especially since it's basically impossible to search for all instances of direct rendering (which is part of why we want to remove it). So we will leave it in for a while so it doesn't break anyone. For now it's not mentioned in the docs anymore so new users wouldn't use it.

 In a later update we'll start logging warnings when it's used and try to make it easy to transition away from it.

**Comment by DawidWraga** [comment]
Next.js examples, especially using React Server Components and the new App directory, would be much appreciated. 

Also, it is not clear where these should be placed eg inside use effect? in layout?
```ts
enableReactUse();
enableReactComponents();
```

**Comment by jmeistrich** [comment]
> Also, it is not clear where these should be placed eg inside use effect? in layout?

They only need to be called once, and it just has to be before their effects are used. It should generally be at the top of a file that is the entry point of your app or is imported everywhere, or it could be at the top of a global state file.

 In our Next apps we have a common `Layout` component that's used on all pages so we put it at the top of that file.

I'll add more about that to the configuring page.

**Comment by DawidWraga** [comment]
Thanks for your reply.

Also, the docs mention the babel plugin multiple times but I can't find any link to it or docs on how to implement it. Any guidance, particularly for those using using next.js, would be much appreciated.

**Comment by jmeistrich** [comment]
@DawidWraga Thanks for catching that! I accidentally deleted the babel installation section while shuffling docs around. Fixed!

**Comment by minorgod** [comment]
> Next.js examples, especially using React Server Components and the new App directory, would be much appreciated.
> 
> Also, it is not clear where these should be placed eg inside use effect? in layout?
> 
> ```ts
> enableReactUse();
> enableReactComponents();
> ```
> 
> I really appreciate your hard work! please keep it up! Thanks

Here's a repo from [@jherr](https://github.com/jherr) (Jack Herrington) showing legend-state with Next 13 and app directory. 
https://github.com/jherr/legend-state-next/tree/main

BTW, if you haven't seen any of Jack Herr's YouTube tutorials for React/NextJS, seriously go check them out. They are short and super informative! In fact that's how I found out about legend-state in the first place. :)

**Comment by FranciscoKloganB** [comment]
I personally would really enjoy a section outline best practises for testing with `legend-state`, particularly, on how to avoid concurrency issues on your tests and test isolation. For now, the only thing I am using `legend-state` stores is for the `AuthStore`. Everything else is react query based.

However, if on the one hand `legend-state` gives me the freedom to create different stores with different patterns as the project evolves, whereas other libs, for example, `XState` would force me into a pattern even in situations where the state machine pattern is not ideal. I want to be able to right comprehensive tests for the actions I have created for my stores. Unfortunatly, my "observables" know-how is limited and the documentation does not mention testing at all, to the point where I am considering choosing something more complex, with bigger bundle sizes, just for this particular reason.

To get me going, I would just need to know the answer to questions like:

- Running tests on the store through dummy test components that allow to test store results through actions and state display
- Running tests on the store without react components (pure js)
- Ensuring store state is not shared across tests running in different/same files.

And any other tips you guys might have.

**Comment by jmeistrich** [comment]
@FranciscoKloganB Maybe a good start is to look at our internal tests: https://github.com/LegendApp/legend-state/tree/main/tests

Are there other documentations that do a good job of describing testing that I could look at for inspiration?

Dummy test components is interesting. You could have a component that just subscribes to the root of a store and displays the whole object?

**Comment by FranciscoKloganB** [comment]
Hello @jmeistrich thanks for taking interesting.

Yes. Dummy components is what I am using to test `route` changes to avoid mocking `react-router-dom` stuff like `history` and `location`, but also to test some custom hooks for `password` visibility. It's a decent pattern and recommend by at least a few players including `testing-library`.

Here are some examples I found that could be used as inspiration:

- https://jotai.org/docs/guides/testing
- https://xstate.js.org/docs/guides/testing.html
- https://tkdodo.eu/blog/testing-react-query#queryclientprovider
  - This one mentions the importance of isolation of the react query provider test to test, I assume the concern could be shared also to a global store that is simply created like this `export const auth$ = observable({ stuff })`. From test file to test file, the `auth$` context would be recreated, but from `test to test` within a file, I do not think so.

**Comment by Wulfre** [comment]
> Next.js examples, especially using React Server Components and the new App directory, would be much appreciated.
> 
> Also, it is not clear where these should be placed eg inside use effect? in layout?
> 
> ```ts
> enableReactUse();
> enableReactComponents();
> ```
> 
> I really appreciate your hard work! please keep it up! Thanks

A few months late to this specific point, but I created a wrapper component that I use in my main layout.tsx so that I can still pass server components as children and not need to make the whole app run on the client. Pardon my style, I believe it's a bit out of the norm for react.

```ts
// LegendStateProvider.tsx

"use client"

import type { FunctionComponent, ReactNode } from "react"
import { enableReactUse } from "@legendapp/state/config/enableReactUse"

enableReactUse() // enables `use()` method on observables in client components

type Props = {
    children: ReactNode
}

const LegendStateProvider: FunctionComponent<Props> = (props) => props.children

export default LegendStateProvider
```

```ts
// app/layout.tsx

import type { FunctionComponent, ReactNode } from "react"
import LegendStateProvider from "~/LegendStateProvider"
import "~/app/globals.css"

type Props = {
    children: ReactNode
}

const RootLayout: FunctionComponent<Props> = (props) => (
    <html lang={"en"}>
        <body>
            <LegendStateProvider>
                {props.children}
            </LegendStateProvider>
        </body>
    </html>
)

export default RootLayout
```

**Comment by briandipalma** [comment]
Delete `$` from all the documentation, IMO it's just noise that might put people off from reading the code.

**Comment by minorgod** [comment]
I agree with @briandipalma. I found it very confusing seeing the variables named with `$` in various places and wasn't sure if it might even be triggering some functionality that wouldn't work without using that variable naming convention. At the very least there should be some clarification on whether that convention is required for any specific functionality or not.

**Comment by linkb15** [comment]
I would suggest to add Recipes sections where some patterns of usage are popularized opinionated by the community. To help newcomers to try things the right way faster.

Troubleshooting sections would be nice to have as well incase for occurring problems that are faced by many people that can be linked into the error messages as well.

**Comment by alban-ameti** [comment]
Hi,

It would be great to see in the docs the rules for syncing items. For example:

1) Item added on client 1 while offline, item added on client 2 while offline, both come online what is the state now?
2) What happens when two clients edit same item while offline what is the result of that?

It would be great if there was a section dedicated to all the different scenarios and how they get resolved.
